In [107]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [109]:
df=pd.read_csv(r"C:\Users\ashwitha\Downloads\archive (18).zip")

In [111]:
df.head()

customer_id  credit_score country  gender  age  tenure    balance  \
0     15634602           619  France  Female   42       2       0.00   
1     15647311           608   Spain  Female   41       1   83807.86   
2     15619304           502  France  Female   42       8  159660.80   
3     15701354           699  France  Female   39       1       0.00   
4     15737888           850   Spain  Female   43       2  125510.82   

   products_number  credit_card  active_member  estimated_salary  churn  
0                1            1              1         101348.88      1  
1                1            0              1         112542.58      0  
2                3            1              0         113931.57      1  
3                2            0              0          93826.63      0  
4                1            1              1          79084.10      0

In [113]:
df.isnull().sum()

customer_id         0
credit_score        0
country             0
gender              0
age                 0
tenure              0
balance             0
products_number     0
credit_card         0
active_member       0
estimated_salary    0
churn               0
dtype: int64

In [115]:
df.duplicated().sum()

0

In [117]:
df.dtypes

customer_id           int64
credit_score          int64
country              object
gender               object
age                   int64
tenure                int64
balance             float64
products_number       int64
credit_card           int64
active_member         int64
estimated_salary    float64
churn                 int64
dtype: object

In [119]:
df.columns

Index(['customer_id', 'credit_score', 'country', 'gender', 'age', 'tenure',
       'balance', 'products_number', 'credit_card', 'active_member',
       'estimated_salary', 'churn'],
      dtype='object')

In [121]:
cols=["customer_id","gender","age","country"]
df= df.drop(cols,axis=1)

In [123]:
df.head()

credit_score  tenure    balance  products_number  credit_card  \
0           619       2       0.00                1            1   
1           608       1   83807.86                1            0   
2           502       8  159660.80                3            1   
3           699       1       0.00                2            0   
4           850       2  125510.82                1            1   

   active_member  estimated_salary  churn  
0              1         101348.88      1  
1              1         112542.58      0  
2              0         113931.57      1  
3              0          93826.63      0  
4              1          79084.10      0

In [125]:
df.columns

Index(['credit_score', 'tenure', 'balance', 'products_number', 'credit_card',
       'active_member', 'estimated_salary', 'churn'],
      dtype='object')

In [127]:
df.dtypes

credit_score          int64
tenure                int64
balance             float64
products_number       int64
credit_card           int64
active_member         int64
estimated_salary    float64
churn                 int64
dtype: object

In [129]:
from sklearn.model_selection import train_test_split

In [131]:
x = df.drop('churn',axis=1)
y = df[['churn']]

In [133]:
X_train,x_test,Y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

In [135]:
X_train

credit_score  tenure    balance  products_number  credit_card  \
9254           686       6       0.00                2            1   
1561           632       4  119624.60                2            1   
1670           559       3  114739.92                1            1   
6087           561       9  135637.00                1            1   
6669           517       9  142147.32                1            0   
...            ...     ...        ...              ...          ...   
5734           768       8   69712.74                1            1   
5191           682       1       0.00                1            1   
5390           735       1       0.00                3            0   
860            667       8  190227.46                1            1   
7270           697       1  147910.30                1            1   

      active_member  estimated_salary  
9254              1         179093.26  
1561              1         195978.86  
1670              0          85891.02  
6087              0         153080.40  
6669              0          39488.04  
...             ...               ...  
5734              1          69381.05  
5191              1            706.50  
5390              0          92220.12  
860               0          97508.04  
7270              1          53581.14  

[8000 rows x 7 columns]

In [137]:
Y_train

churn
9254      0
1561      0
1670      1
6087      1
6669      1
...     ...
5734      0
5191      0
5390      1
860       1
7270      0

[8000 rows x 1 columns]

In [155]:
X_train.shape

(8000, 7)

In [157]:
Y_train.shape

(8000, 1)

In [139]:
from sklearn.preprocessing import OneHotEncoder

In [141]:
san=OneHotEncoder(drop="first",sparse_output=False)

In [143]:
df.dtypes

credit_score          int64
tenure                int64
balance             float64
products_number       int64
credit_card           int64
active_member         int64
estimated_salary    float64
churn                 int64
dtype: object

In [147]:
df.dtypes

credit_score          int64
tenure                int64
balance             float64
products_number       int64
credit_card           int64
active_member         int64
estimated_salary    float64
churn                 int64
dtype: object

In [149]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

In [161]:
model=LogisticRegression(max_iter=1000,class_weight='balanced')
model.fit(X_train,Y_train)
y_pred=model.predict(x_test)
print("accuracy:",accuracy_score(y_test,y_pred))
print("confusion matrix:\n",confusion_matrix(y_test,y_pred))
print("classification_report:\n",classification_report(y_test,y_pred))

C:\Users\ashwitha\anaconda3\Lib\site-packages\sklearn\utils\validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


accuracy: 0.65
confusion matrix:
 [[1075  532]
 [ 168  225]]
classification_report:
               precision    recall  f1-score   support

           0       0.86      0.67      0.75      1607
           1       0.30      0.57      0.39       393

    accuracy                           0.65      2000
   macro avg       0.58      0.62      0.57      2000
weighted avg       0.75      0.65      0.68      2000



C:\Users\ashwitha\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [163]:
from sklearn.preprocessing import StandardScaler

In [171]:
scaler=StandardScaler()
X_train=scaler.fit_transform(X_train)
x_test=scaler.transform(x_test)
model=LogisticRegression(max_iter=5000,class_weight="balanced")
model.fit(X_train,Y_train)
y_pred=model.predict(x_test)
print("accuracy:",accuracy_score(y_test,y_pred))
print("confusion matrix:\n",confusion_matrix(y_test,y_pred))
print("classification_report:\n",classification_report(y_test,y_pred))
y_prob=model.predict_proba(x_test)[:,-1]



accuracy: 0.649
confusion matrix:
 [[1073  534]
 [ 168  225]]
classification_report:
               precision    recall  f1-score   support

           0       0.86      0.67      0.75      1607
           1       0.30      0.57      0.39       393

    accuracy                           0.65      2000
   macro avg       0.58      0.62      0.57      2000
weighted avg       0.75      0.65      0.68      2000



C:\Users\ashwitha\anaconda3\Lib\site-packages\sklearn\utils\validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [173]:
threshold=0.6
y_pred=(y_prob >=threshold).astype(int)
print("threshold:",threshold)
print("accuracy:",accuracy_score(y_test,y_pred))
print("confusion matrix:\n",confusion_matrix(y_test,y_pred))
print("classification_report:\n",classification_report(y_test,y_pred))


threshold: 0.6
accuracy: 0.7075
confusion matrix:
 [[1272  335]
 [ 250  143]]
classification_report:
               precision    recall  f1-score   support

           0       0.84      0.79      0.81      1607
           1       0.30      0.36      0.33       393

    accuracy                           0.71      2000
   macro avg       0.57      0.58      0.57      2000
weighted avg       0.73      0.71      0.72      2000



In [175]:
threshold=0.5
y_pred=(y_prob >=threshold).astype(int)
print("threshold:",threshold)
print("accuracy:",accuracy_score(y_test,y_pred))
print("confusion matrix:\n",confusion_matrix(y_test,y_pred))
print("classification_report:\n",classification_report(y_test,y_pred))


threshold: 0.5
accuracy: 0.649
confusion matrix:
 [[1073  534]
 [ 168  225]]
classification_report:
               precision    recall  f1-score   support

           0       0.86      0.67      0.75      1607
           1       0.30      0.57      0.39       393

    accuracy                           0.65      2000
   macro avg       0.58      0.62      0.57      2000
weighted avg       0.75      0.65      0.68      2000



In [179]:
for t in[0.4,0.5,0.6]:
    y_pred=(y_prob >= t).astype(int)
    print("\nthreshold:",t)
    print(confusion_matrix(y_test,y_pred))
    print(accuracy_score(y_test,y_pred))



threshold: 0.4
[[ 411 1196]
 [  41  352]]
0.3815

threshold: 0.5
[[1073  534]
 [ 168  225]]
0.649

threshold: 0.6
[[1272  335]
 [ 250  143]]
0.7075


In [181]:
for t in[0.5,0.55,0.6]:
    y_pred=(y_prob >= t).astype(int)
    print("\nthreshold:",t)
    print(confusion_matrix(y_test,y_pred))
    print(accuracy_score(y_test,y_pred))


threshold: 0.5
[[1073  534]
 [ 168  225]]
0.649

threshold: 0.55
[[1204  403]
 [ 211  182]]
0.693

threshold: 0.6
[[1272  335]
 [ 250  143]]
0.7075


In [183]:
final_threshold=0.55
y_prob=model.predict_proba(x_test)[:,-1]
y_pred_final=(y_prob >=final_threshold).astype(int)


In [185]:
print("threshold:",final_threshold)
print("accuracy:",accuracy_score(y_test,y_pred))
print("confusion matrix:\n",confusion_matrix(y_test,y_pred))
print("classification_report:\n",classification_report(y_test,y_pred))

threshold: 0.55
accuracy: 0.7075
confusion matrix:
 [[1272  335]
 [ 250  143]]
classification_report:
               precision    recall  f1-score   support

           0       0.84      0.79      0.81      1607
           1       0.30      0.36      0.33       393

    accuracy                           0.71      2000
   macro avg       0.57      0.58      0.57      2000
weighted avg       0.73      0.71      0.72      2000



 ## Logistic Regression Model – Final Results (Threshold = 0.55)

The churn dataset is imbalanced (more non-churn customers than churn customers).
Initially, logistic regression predicted mostly the majority class (non-churn), so class balancing and threshold tuning were applied.

The final model was trained using Logistic Regression with class_weight='balanced' and evaluated using a tuned threshold of 0.55.|

## Final Metrics

Accuracy: 0.7075

Confusion Matrix:

[
1272
	
335


250
	
143
]
[
1272
250
	​

335
143
	​

]
## Interpretation

1272 customers were correctly predicted as not churn (True Negatives).

335 customers were incorrectly predicted as churn (False Positives).

250 churn customers were missed (False Negatives).

143 churn customers were correctly detected (True Positives).

Class-1 (Churn) Performance

 ## Precision: 0.30

 ## Recall: 0.36

## F1-Score: 0.33

## This threshold (0.55) provides a better balance between reducing false positives and detecting churn customers compared to the default threshold.

In [190]:
import pickle

final_threshold = 0.55

model_data = {
    "model": model,
    "threshold": final_threshold
}

with open("logistic_churn_model.pkl", "wb") as file:
    pickle.dump(model_data, file)

print("Model + threshold saved successfully!")


Model + threshold saved successfully!
